controller/resource.go

```diff
@@ -0,0 +1,240 @@
+		}
+		for _, task := range tasks.Tasks {
+			resources = append(resources, &Task{
+				SecretsManagerClient: t.SecretsManagerClient,
+				ConsulClient:         t.ConsulClient,
+				Cluster:              t.Cluster,
+				Log:                  t.Log,
+				SecretPrefix:         t.SecretPrefix,
+				Task:                 task,
+			})
+		}
+		if nextToken == nil {
+			break
+		}
+	}
+	return resources, nil
+}
+
+type Task struct {
+	SecretsManagerClient secretsmanageriface.SecretsManagerAPI
+	ConsulClient         *api.Client
+
+	Cluster      string
+	SecretPrefix string
+	Task         *ecs.Task
+
+	Log hclog.Logger
+}
+
+// ID returns Task definition ARN or error if it cannot be determined from the Task.
+func (t *Task) ID() (string, error) {
+	// This should never be the case, but we are checking it anyway.
+	if t.Task.TaskDefinitionArn == nil {
+		return "", fmt.Errorf("cannot determine ID: task definition ARN is nil")
+	}
+
+	return *t.Task.TaskDefinitionArn, nil
+}
+
+// tokenSecretJSON is the struct that represents JSON of the token secrets
+// stored in Secrets Manager.
+type tokenSecretJSON struct {
+	AccessorID string `json:"accessor_id"`
+	Token      string `json:"token"`
+}
+
+// Upsert creates a token for the task if one doesn't already exist
+// and updates the secret with the contents of the token.
+func (t *Task) Upsert() error {
+	serviceName, err := t.parseFamilyNameFromTaskDefinitionARN(*t.Task.TaskDefinitionArn)
+	if err != nil {
+		return fmt.Errorf("could not determine service name: %w", err)
+	}
+
+	meshTask := tagValue(t.Task.Tags, meshTag) == "true"
+	if !meshTask {
+		t.Log.Info("skipping non-mesh task", "id", serviceName)
+		return nil
+	}
+
+	secretName := fmt.Sprintf("%s-%s", t.SecretPrefix, serviceName)
+
+	// Get current secret from AWS.
+	currSecretValue, err := t.SecretsManagerClient.GetSecretValue(&secretsmanager.GetSecretValueInput{SecretId: aws.String(secretName)})
+	if err != nil {
+		return fmt.Errorf("retrieving secret: %w", err)
+	}
+	var currSecret tokenSecretJSON
+	err = json.Unmarshal([]byte(*currSecretValue.SecretString), &currSecret)
+	if err != nil {
+		return fmt.Errorf("unmarshalling secret value: %w", err)
+	}
+
+	var currToken *api.ACLToken
+	// If we already have an accessor ID, we'll check if this token exists in Consul first.
+	// We don't care if the token value is empty or not in this case.
+	// If token value is empty, then it's an empty secret, and we should update it with the token.
+	// If token value is non-empty it indicates that something is corrupted, and we should update the token.
+	if currSecret.AccessorID != "" {
+		// Read the token with this Accessor ID from Consul.
+		currToken, _, err = t.ConsulClient.ACL().TokenRead(currSecret.AccessorID, nil)
+
+		if err != nil && !isACLNotFoundError(err) {
+			return fmt.Errorf("reading existing token: %w", err)
+		}
+	}
+
+	// If there is already a token for this service in Consul, exit early.
+	if currToken != nil {
```

Comment:
Should we log something for this case?

Reply:
I guess you've only got error logs here. Should we have some info logs letting users know that a token was creatd?